server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java

```diff
@@ -182,60 +181,85 @@ private static void advance(final BytesRef ref, final int length) {
     }
 
     /**
-     * Instead of adding the complexity of {@link InputStream#reset()} etc to the actual impl
-     * this wrapper builds it on top of the BytesReferenceStreamInput which is much simpler
-     * that way.
+     * Mark-supporting {@link StreamInput} over all the bytes in a {@link BytesReference}.
      */
     private static final class MarkSupportingStreamInputWrapper extends StreamInput {
-        // can't use FilterStreamInput it needs to reset the delegate
         private final BytesReference reference;
-        private BytesReferenceStreamInput input;
+        private int offset = 0;
         private int mark = 0;
 
-        private MarkSupportingStreamInputWrapper(BytesReference reference) throws IOException {
+        private MarkSupportingStreamInputWrapper(BytesReference reference) {
             this.reference = reference;
-            this.input = new BytesReferenceStreamInput(reference.iterator(), reference.length());
         }
 
         @Override
         public byte readByte() throws IOException {
-            return input.readByte();
+            if (offset >= reference.length()) {
+                throw new EOFException();
+            }
+            return reference.get(offset++);
         }
 
         @Override
-        public void readBytes(byte[] b, int offset, int len) throws IOException {
-            input.readBytes(b, offset, len);
+        public int readInt() {
```

Comment:
We could potentially get a pretty serious speedup in a follow-up to this one by moving more deserialization to the `BytesReference` implementations as done for `int` here. For things like deserializing a `String` we could completely remove the need for copying any bytes to an intermediate buffer and just and save tons of bounds checks.

Reply:
I'm definitely worried about losing the `EOFException` here.

Reply:
I'm not I think (it's the opposite of the other spot in terms sorta). This will throw a `RuntimeException` instead of an `EOFException` now in some spots, which should rather than hide make serialization bugs visible. We never actually seem to catch `EOFException` specifically in serialization logic so whenever we throw here it's a bug anyway IMO.